cylc/flow/task_job_mgr.py

```diff
@@ -997,44 +1003,74 @@ def _set_retry_timers(
+        * Dummy: Pre-submission preparation (removing task scripts content)
+          before returning to live pathway.
+        * Live: return to main submission pathway without doing anything.
+
+        Returns:
+            lively_tasks:
+                A list of tasks which require subsequent
+                processing **as if** they were live mode tasks.
+                (This includes live and dummy mode tasks)
+            nonlive_tasks:
+                A list of tasks which require no further processing
+                because their apparent execution is done entirely inside
+                the scheduler. (This includes skip and simulation mode tasks).
+        """
+        lively_tasks: 'List[TaskProxy]' = []
+        nonlive_tasks: 'List[TaskProxy]' = []
         now = time()
-        now_str = get_time_string_from_unix_time(now)
+        now = (now, get_time_string_from_unix_time(now))
+
         for itask in itasks:
-            # Handle broadcasts
+            # Get task config with broadcasts applied:
             rtconfig = self.task_events_mgr.broadcast_mgr.get_updated_rtconfig(
                 itask)
 
-            itask.summary['started_time'] = now
-            self._set_retry_timers(itask, rtconfig)
-            itask.mode_settings = ModeSettings(
-                itask,
-                self.workflow_db_mgr,
-                rtconfig
-            )
-
-            itask.waiting_on_job_prep = False
-            itask.submit_num += 1
+            # Apply task run mode
+            if workflow_run_mode in RunMode.NON_OVERRIDABLE_MODES.value:
+                # Task run mode cannot override workflow run-mode sim or dummy:
+                run_mode = workflow_run_mode
+            else:
+                # If workflow mode is skip or live and task mode is set,
+                # override workflow mode, else use workflow mode.
+                run_mode = rtconfig.get('run mode', None) or workflow_run_mode
+            # Store the run mode of the this submission:
+            itask.run_mode = run_mode
+
+            # Submit nonlive tasks, or add live-like (live or dummy)
+            # tasks to list of tasks to put through live
+            # submission pipeline - We decide based on the output
+            # of the submit method:
+            is_nonlive = False
+            if run_mode == RunMode.DUMMY.value:
+                is_nonlive = dummy_submit_task_job(
+                    self, itask, rtconfig, workflow, now)
+            elif run_mode == RunMode.SIMULATION.value:
+                is_nonlive = simulation_submit_task_job(
+                    self, itask, rtconfig, workflow, now)
+            elif run_mode == RunMode.SKIP.value:
+                is_nonlive = skip_submit_task_job(
+                    self, itask, rtconfig, now)
+
+            # Assign task to list:
+            if is_nonlive:
+                nonlive_tasks.append(itask)
+            else:
+                lively_tasks.append(itask)
```

Comment:
> It would make sense to import these modules dynamically as few schedulers will use more than 1 of them.

It's taken me a few minutes to satisfy myself that this is safe. I didn't like the idea of having an import every time we ran a task, but it looks like Python will handle this. Otherwise, lovely.

Reply:
Here's another example of dynamic imports in the Cylc codebase:

https://github.com/cylc/cylc-flow/blob/bf02ddb8197cddfba8145f9d4ac9ea076e685537/cylc/flow/scheduler_cli.py#L423

Note, Python's import interface is doing something like this under the hood:

```python
def __import__(self, module_name):
    try:
        return sys.modules[module_name]
    except:
        mod = _import(module_name)
        sys.modules[module_name] = module
```

I.E. it doesn't actually re-import the module, it just returns the cached instance.

Which is why something like this works:

```python
>>> import cylc.flow.flags as flags
>>> flags.verbosity
0
>>> flags.verbosity = 1
>>> flags.verbosity
1
>>> import cylc.flow.flags as flags
>>> flags.verbosity
1
```